tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1173,183 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
     return 0;
 }
 
-static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+/* This function returns the client index for a cipher *match. If a matching cipher could not be found, then -1 is returned. */
+static int s2n_wire_ciphers_has_server_cipher_at(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
 {
+    int negative_one = -1;
     for (uint32_t i = 0; i < count; i++) {
         const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
 
-        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
-            return 1;
+        if (s2n_constant_time_equals(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
         }
     }
 
-    return 0;
+    return negative_one;
+}
+
+static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    int server_index = s2n_wire_ciphers_has_server_cipher_at(match, wire, count, cipher_suite_len);
+    if (server_index < 0) {
+        return 0;
+    }
+    return 1;
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+        /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+        if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+            return false;
+        }
+
+        /* If connection is for SSLv3, use SSLv3 version of suites */
+        if (conn->client_protocol_version == S2N_SSLv3) {
```

Comment:
You have this if-statement in two places in your code. Why?

Reply:
Here, it is required because the validation needs to be done on the sslv3 version of the cipher suite (if it exists). This is done again in `s2n_set_cipher_as_server` because the function just returns the index of the cipher, not the actual cipher itself.